File: 0494-Target Sum/0494-Target Sum.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int S) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if (sum + S < 0 || S > sum)
        {
            return 0;
        }
        for (int i = 0; i < nums.size(); ++i)
        {
            nums[i] *= 2;
        }
        
        vector<int> dp(1 + sum + S);
        dp[0] = 1;
        for (int i = 0; i < nums.size(); ++i)
        {
            for (int j = sum + S; j >= nums[i]; --j)
            {
                dp[j] += dp[j - nums[i]];
            }
        }
        
        return dp[sum + S];
    }
};

File: 0401-0500/0494-Target Sum/0494-Target Sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], S: int) -> int:
        total = sum(nums)
        if S > total or S < -total:
            return 0
        for i in range(len(nums)):
            nums[i] *= 2
        dp = [0] * (1 + S + total)
        dp[0] = 1
        for i in range(len(nums)):
            for j in range(S + total, nums[i] - 1, -1):
                dp[j] += dp[j - nums[i]]
        return dp[S + total]
